stakingModel/contracts/SimpleStaking.sol

```diff
@@ -0,0 +1,232 @@
+
+    ERC20 token;
+    ERC20 iToken;
+
+    uint256 public blockInterval;
+    uint256 lastUBICollection;
+    uint256 public totalStaked = 0;
+
+    //how much of the generated interest is donated, meaning no G$ is expected in compensation, 1 in mil precision.
+    //100% for phase0 POC
+    uint32 public avgInterestDonatedRatio = 1e6;
+
+    address public fundManager;
+
+    modifier onlyFundManager {
+        require(msg.sender == fundManager, "Only FundManager can call this method");
+        _;
+    }
+
+
+    constructor(
+        address _token,
+        address _iToken,
+        address _fundManager,
+        uint256 _blockInterval,
+        Avatar _avatar,
+        Identity _identity
+    ) public FeelessScheme(_identity, _avatar) {
+        token = ERC20(_token);
+        iToken = ERC20(_iToken);
+        blockInterval = _blockInterval;
+        lastUBICollection = block.number.div(blockInterval);
+        fundManager = _fundManager;
+        addPauser(address(avatar));
+    }
+
+    /**
+    @dev allow the DAO to change the fund manager contract
+    @param _fundManager address of the new contract
+    */
+    function setFundManager(address _fundManager) public onlyAvatar {
+        fundManager = _fundManager;
+    }
+
+    /**
+     * @dev stake some Token
+     * @param amount of Token to stake
+     */
+    function stake(uint256 amount) external whenNotPaused {
+        
+        require(amount > 0, "You need to stake a positive token amount");
+        require(
+            token.transferFrom(msg.sender, address(this), amount),
+            "transferFrom failed, make sure you approved token transfer"
+        );
+
+        // approve the transfer to defi protocol
+        token.approve(address(iToken), amount);
+        mint(amount); //mint iToken
+
+        Staker storage staker = stakers[msg.sender];
+        staker.stakedToken = staker.stakedToken.add(amount);
+        staker.lastStake = block.number;
+        totalStaked = totalStaked.add(amount);
+        emit Staked(msg.sender, address(token), amount);
+    }
+
+    /**
+     * @dev withdraw all staked Token
+     */
+    function withdrawStake() external {
+        Staker storage staker = stakers[msg.sender];
+        require(staker.stakedToken > 0, "No DAI staked");
+        uint256 tokenWithdraw = staker.stakedToken;
+        uint256 tokenActual = token.balanceOf(address(this));
+        if (tokenActual < tokenWithdraw) {
+            tokenWithdraw = tokenActual;
+        }
+        redeem(tokenWithdraw);
+        staker.stakedToken = staker.stakedToken.sub(tokenWithdraw); // update balance before transfer to prevent re-entry
+        totalStaked = totalStaked.sub(tokenWithdraw);
+        require(token.transfer(msg.sender, tokenWithdraw), "withdraw transfer failed");
+        emit StakeWithdraw(msg.sender, address(token), tokenWithdraw, tokenActual);
+    }
+
+    function currentTokenWorth() public view returns (uint256) {
+        uint256 er = exchangeRate();
```

Comment:
i dont think its possible. to use exchangeRate that will be defined in a child contract ie in GoodCompoundStaking

Reply:
Hi @sirpy , I believe this should work. My approach is that we only deploy child contracts like GoodCompoundStaking contract and never deploy parent contract SimpleStaking directly. Tomorrow if we wish to introduce a new DeFi protocol eg:- Aave, we can very quickly deploy GoodAaveStaking reusing the entire code of SimpleStaking and only writing functions specific to Aave protocol. Let  me know if this makes sense.